### PR TITLE
Attenuation Effect

### PR DESCRIPTION
**Pull request by romulogcerqueira**
Include attenuation effect directly on GPU calculus of sound echo intensity and unitary tests.

Attenuation Coeff. = Borid Acid Contribution + Magnesium Sulphate Contribution + Freshwater Contribution

@trocolit2, please have a look at this PR!

### CODE REVIEW

File: src/Tools.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                                    const double depth,
+                                    const double salinity,
+                                    const double acidity) {
+
+    // borid acid and magnesium sulphate relaxation frequencies (in kHz)
+    double f1 = 0.78 * pow(salinity / 35, 0.5) * exp(temperature / 26);
+    double f2 = 42 * exp(temperature / 17);
+
+    // borid acid contribution
+    double borid = 0.106 * ((f1 * frequency * frequency) / (frequency * frequency + f1 * f1)) * exp((acidity - 8) / 0.56);
```
Comment (trocolit2) [inline src/Tools.cpp]: Just a style code suggestion. 
Even rock style allows us write long line codes (200 chars), I advice you rewrite and reduce this line size to 80 chars length. It is better to read.

File: src/Tools.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    // magnesium sulphate contribuion
+    double magnesium = 0.52 * (1 + temperature / 43) * (salinity / 35)
+                        * ((f2 * frequency * frequency) / (frequency * frequency + f2 * f2)) * exp(-depth / 6000);
```
Comment (trocolit2) [inline src/Tools.cpp]: In your code, you use many ` frequency * frequency `. I could compute at the begin of the method and keep  the result in other variable like : 
`double frequency2 = frequency * frequency;`

File: test/Attenuation_test.cpp [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    double depth = 1;           // meters
+    double salinity = 35;       // ppt
+    double acidity = 8.1;       // pH
+
+    double attenuationCoeff = underwaterSignalAttenuation(frequency, temperature, depth, salinity, acidity);
+    BOOST_CHECK_CLOSE(attenuationCoeff, 0.0247, 3);
+}
+
+// simple sonar plot
+void plotSonarTest(cv::Mat3f image, double maxRange, double maxAngleX) {
```
Comment (trocolit2) [inline test/Attenuation_test.cpp]: It is better make a new CPP file with all methods required to support tests.

File: test/Attenuation_test.cpp [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    makeSimpleScene(root);
+
+    // display the same with and without underwater acoustic attenuation
+    for (uint i = 0; i < eyes.size(); ++i) {
+        cv::Mat raw_shader = computeNormalDepthMap(root, height, maxRange, fovX, fovY, 0, eyes[i], centers[i], ups[i]);
+        cv::Mat att_shader = computeNormalDepthMap(root, height, maxRange, fovX, fovY, attenuationCoeff, eyes[i], centers[i], ups[i]);
+        plotSonarTest(raw_shader, maxRange, fovX * 0.5);
+        plotSonarTest(att_shader, maxRange, fovX * 0.5);
+  }
+}
```
Comment (trocolit2) [inline test/Attenuation_test.cpp]: How do you test the code are correctly?
You just plot the final image and check visually?
IMO, you should choose some reference points in final image (sonar plot or raw shader image) and compare these points with reference value. This way you can track and sense if any code change gives the same expected result.
See the normalDepthMap_test.cpp.

File: src/Tools.cpp (hunk 2 above)
Comment (romulogcerqueira) [reply src/Tools.cpp]: Good practice!

File: test/Attenuation_test.cpp (hunk 3 above)
Comment (romulogcerqueira) [reply test/Attenuation_test.cpp]: Ok.

File: test/Attenuation_test.cpp (hunk 4 above)
Comment (romulogcerqueira) [reply test/Attenuation_test.cpp]: Ok. I will add this check.

**Comment by romulogcerqueira** [comment]
Implemented suggestions. Please have a look again, @trocolit2

**Review by trocolit2 — APPROVED** [review]

**Pull request merged**